### DISCUSSION TITLE
Smarter cursor placement around inlay hints

### DISCUSSION THREAD

**Discussion by mydumpfire** [helix-editor/helix#10001] [General]
Cursor placement and character insertion near inlay hints currently feels a bit janky:

[helix_hint.webm](https://github.com/helix-editor/helix/assets/134539278/af241e8c-d4de-4dc7-b171-48296a0dfe94)

This is how it looks in VS Code:

[vscode_hint.webm](https://github.com/helix-editor/helix/assets/134539278/67839ab8-84b4-4a73-8f27-318fc557dff8)

Neovim also deals with this pretty well:

https://github.com/helix-editor/helix/assets/134539278/72ae5b1b-28bf-4ee1-b2f3-db60029877ee

I'm not sure what they do under the hood, but it feels like they always do the "right" thing, positioning your cursor where it makes the most sense in the current context.

**Comment by RossSmyth** [comment]
For exploration it seems this is a PR that touched it https://github.com/microsoft/vscode/pull/128140
VSCode's structure is very complex for my small embedded software brain so I have no idea if there are newer PRs.

**Comment by pascalkuthe** [comment]
Our cursor semantics are fundamentally different than vscodes. Vscode has a cursor between character while we have cursors on characters. Our implementation is consistent with that behavior. 

The problem is mostly that you are using bar cursor which are just a visual gimmick and don't reflect the actual cursor semantics which are always block cursor based.

**Reply by mydumpfire** [reply]
I don't think it's *just* a gimmick problem. For example notice how the inlay hint gets sandwiched in the middle of the currently inputted text. I don't think this should ever happen, even with block-based cursors.

Neovim is also an editor with a block-based cursor, but it doesn't have this problem at all:

https://github.com/helix-editor/helix/assets/134539278/72ae5b1b-28bf-4ee1-b2f3-db60029877ee



Can this be re-opened as an issue?

**Reply by The-Lost-Light** [reply]
I agree that. The cursor around inlay hint looks weird.

**Reply by tp86** [reply]
I think the problem is especially visible in append mode.

In append mode cursor is NOT part of selection. Upon entering append mode it is moved one character to the right, on leaving, moved one character to left. Inlay hint just after cursor in normal mode causes cursor to be displayed after hint, while characters are inserted before hint, which is very confusing with long hints.
